### PR TITLE
D42 14245 - Modify FreshService Script to map OS information

### PR DESCRIPTION
**Pull request by osanchez42**
- added support for variations in os name
- added os version to mapping file

### CODE REVIEW

File: d42_sd_sync.py [hunk 1]
```diff
@@ -59,7 +59,39 @@ def get_asset_type_field(asset_type_fields, map_info):
...
+        }
+
+    for os_key, alternate_os_names in accepted_operating_systems.items():
+        for alternate_name in alternate_os_names:
+            if alternate_name in original_os_name.lower():
+                return os_key.title()
+
+
 def get_map_value_from_device42(source, map_info, b_add=False, asset_type_id=None):
+    # D42-14245
```
Comment (cscaglioned42) [inline d42_sd_sync.py]: Add title of the ticket also:
Modify FreshService Script to map OS information

File: d42_sd_sync.py [hunk 2]
```diff
@@ -59,7 +59,39 @@ def get_asset_type_field(asset_type_fields, map_info):
...
+        'Mac': ['mac', 'os x', 'macos', 'ios'],
+        'Solaris': ['solaris'],
+        'Aix': ['aix']
+        }
+
+    for os_key, alternate_os_names in accepted_operating_systems.items():
+        for alternate_name in alternate_os_names:
+            if alternate_name in original_os_name.lower():
+                return os_key.title()
+
```
Comment (cscaglioned42) [inline d42_sd_sync.py]: Should probably have a return None at the end of the function just in case the OS name is not found.

**Comment by osanchez42** [comment]
@cscaglioned42 requested changes made, please let me know if there is anything else

File: mapping.xml.sample [hunk 3]
```diff
@@ -9,14 +9,14 @@
...
...w_device_v1.cpupower END as cpuspeed, case WHEN view_device_v1.ram_size_type = 'MB' and view_device_v1.ram is not null  THEN view_device_v1.ram / 1024 WHEN view_device_v1.ram_size_type = 'TB' and view_device_v1.ram is not null  THEN view_device_v1.ram * 1024 ELSE view_device_v1.ram END as ramsize, network.ip_address, network.hwaddress :: macaddr from view_device_v1 left join view_hardware_v1 on view_device_v1.hardware_fk = view_hardware_v1.hardware_pk left join  (select view_ipaddress_v1.device_fk,  max(view_ipaddress_v1.ip_address) as ip_address,  max(view_netport_v1.hwaddress) as hwaddress from view_ipaddress_v1 left join view_netport_v1 on view_ipaddress_v1.netport_fk = view_netport_v1.netport_pk where view_netport_v1.hwaddress is null or (view_netport_v1.hwaddress is not null and LENGTH(view_netport_v1.hwaddress)=12) group by view_ipaddress_v1.device_fk) network on view_device_v1.device_pk=network.device_fk left join (select device_fk, sum(cost) as cost from view_purchaselineitem_v1 left join view_purchaselineitems_to_devices_v1 on purchaselineitem_fk = purchaselineitem_pk group by device_fk) t_cost on t_cost.device_fk=device_pk left join view_vendor_v1 on vendor_pk=vendor_fk "
```
Comment (cscaglioned42) [inline mapping.xml.sample]: I think you can get rid of this and then just map os_name to the OS and OS version fields in Freshservice.

File: mapping.xml.sample [hunk 4]
```diff
@@ -54,16 +54,18 @@
...
-                        <item key="Aix" value="Aix"/>
+                        <item key="linux, ubuntu, centos, oracle, freebsd, suse, debian, fedora, opensuse" value="Linux"/>
+                        <item key="windows, microsoft" value="Windows"/>
+                        <item key="mac, os x, macos, ios" value="Mac"/>
+                        <item key="solaris" value="Solaris"/>
+                        <item key="aix" value="Aix"/>
                     </value-mapping>
                 </field>
+
+                <field resource="os_name_2" source-type="string" target="os_version" target-type="string" target-header="Computer" not-null="true"/>
```
Comment (cscaglioned42) [inline mapping.xml.sample]: Should be able to map os_name to os_version.

File: d42_sd_sync.py [hunk 5]
```diff
@@ -77,9 +79,23 @@ def get_map_value_from_device42(source, map_info, b_add=False, asset_type_id=Non
...
                 items = [map_info["value-mapping"]["item"]]
+
+            # D42-14245 - Modify FreshService Script to map OS information
             for item in items:
-                if item["@key"] == d42_value:
-                    d42_val = item["@value"]
+                try:
+                    if len(item["@key"].split(', ')) > 1:  # if key is list
+                        try:
+                            if d42_value.lower() in item["@key"].split(', '):
```
Comment (cscaglioned42) [inline d42_sd_sync.py]: If d42_value was "Microsoft Windows 10" would that find a match in the array?

**Comment by osanchez42** [comment]
requirements have changed. needs to be recoded to match FS app

**Pull request closed without merging**